### ISSUE TITLE
How to better organize workflow with batches

### ISSUE THREAD

**Issue by benoittgt** [mperham/sidekiq#3522]
Hello

**tl;dr: How to better organize Sidekiq workflow with batches (A->B->C)**

This is not an issue but more advice I'm looking for. I have a simple flow that has 3 steps (A->B->C).
These steps have to run in order. I made a working version that I posted below (it's in a rails app). It's based on [*Really Complex Workflows with Batches*](https://github.com/mperham/sidekiq/wiki/Really-Complex-Workflows-with-Batches)

```ruby
# lib/store_order_flow.rb
class StoreOrderFlow
  def call_batch
    order = Order.last
    overall = Sidekiq::Batch.new
    overall.on(:success, 'FulfillmentCallbacks#shipped', 'oid' => order.id)
    overall.description = "Fulfillment for Order #{order.id}"
    overall.jobs do
      StartWorkflowWorker.perform_async
    end
  end
end
```

```ruby
# app/workers/start_workflow_worker.rb
class StartWorkflowWorker
  include Sidekiq::Worker
  def perform
    batch.jobs do
      step1 = Sidekiq::Batch.new
      step1.on(:success, 'FulfillmentCallbacks#step1_done', 'oid' => 'oid')
      step1.jobs do
        AWorker.perform_async
      end
    end
  end
end
```

```ruby
class FulfillmentCallbacks
  def step1_done(status, options)
    overall = Sidekiq::Batch.new(status.parent_bid)
    overall.jobs do
      step2 = Sidekiq::Batch.new
      step2.on(:success, 'FulfillmentCallbacks#step2_done')
      step2.jobs do
        BWorker.perform_async
      end
    end
  end

  def step2_done(status, options)
    overall = Sidekiq::Batch.new(status.parent_bid)
    overall.jobs do
      step3 = Sidekiq::Batch.new
      step3.jobs do
        CWorker.perform_async
      end
    end
  end

  def shipped(status, options)
    puts "Order #{oid} has shipped!"
  end
end
```

```ruby
# app/workers/a_worker.rb  (same for b and c worker)
class AWorker
  include Sidekiq::Worker
  def perform
    logger.debug 'Inside A'
  end
end
````

So my problem is I found it quite unreadable. It took me some time to understand the nesting of callbacks. For me, the key is the first two part `lib/store_order_flow.rb` and `app/workers/start_workflow_worker.rb` I'm wondering if there is a better way to organize the code of the initial flow and after the three step + ending callback.

Also, I know it's already been discussed [here](https://github.com/mperham/sidekiq/issues/2700#issuecomment-309813886) but I would love to test the flow while mocking the workers. Something like: 

```ruby
describe '#call_worker' do
  before(:all) { Sidekiq::Testing.inline! }

  it 'call the free workers and succeed' do
    expect(AWorker).to receive(:perform_async).and_return('???')
    expect(BWorker).to receive(:perform_async).and_return('???')
    expect(CWorker).to receive(:perform_async).and_return('???')

    StoreOrderFlow.new.call_worker
    # expect(Job.last.status).to eq(Job::STATUS_COMPLETED)
  end

  it 'fails on BWorker' do
    expect(AWorker).to receive(:perform_async).and_return('???')
    expect(BWorker).to receive(:perform_async).and_raise(RuntimeError?)

    StoreOrderFlow.new.call_worker
    # expect(Job.last.status).to eq(Job::STATUS_FAILED)
    # expect(Job.last.step).to eq('b_worker')
  end
  after(:all) { Sidekiq::Testing.fake! }
end
```
I know you are probably correct on _"Executing batches within the test environment would give a false sense of confidence."_ but as you can see above, test could help to better understanding and refactoring my batch workflow.

Last note: Thanks a lot for [*Really Complex Workflows with Batches*](https://github.com/mperham/sidekiq/wiki/Really-Complex-Workflows-with-Batches)

Ruby version: 2.4.1
Sidekiq Pro : 3.3.3

**Comment by mperham** [comment]
I hear you loud and clear.  It's easy to argue that the Batch API is too low-level for easy construction of larger workflows; there is a fair amount of boiler-plate.  I'd like to provide a higher-level API/DSL which makes multi-step workflows simpler to read and construct but it hasn't been a priority so far.  How those workflows are tested is a separate but important consideration.

Just as an easy refactor, you could add helper methods like:

```ruby
  def add_next_step(status, &block)
    overall = Sidekiq::Batch.new(status.parent_bid)
    overall.jobs do
      step = Sidekiq::Batch.new
      step.jobs(&block)
    end
end

  def step2_done(status, options)
    add_next_step(status) do
        CWorker.perform_async
    end
  end
```

**Comment by benoittgt** [comment]
Thanks for your answer. I've implemented a solution without changing that much from the wiki. It's working but I will see how other people in my team will deal with this flow especially newcomers.

Feel free to close the issue if you want.
